File: link_checker.cpp
// link_checker.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <thread>
#include <mutex>
#include <curl/curl.h>
#include <chrono>

using namespace std;

struct Result {
    string url;
    long status;
    bool error;
};

size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    return size * nmemb;
}

long check_url(const string& url, long timeout_sec, bool follow) {
    CURL *curl = curl_easy_init();
    if (!curl) return 0;
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, timeout_sec);
    curl_easy_setopt(curl, CURLOPT_NOBODY, 1L);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, follow ? 1L : 0L);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "LinkChecker/1.0");
    CURLcode res = curl_easy_perform(curl);
    long http_code = 0;
    if (res == CURLE_OK) {
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
    } else {
        http_code = 0;
    }
    curl_easy_cleanup(curl);
    return http_code;
}

vector<string> load_links(const string& filename) {
    vector<string> links;
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Не удалось открыть файл: " << filename << endl;
        return links;
    }
    string line;
    while (getline(file, line)) {
        if (line.empty() || line[0] == '#') continue;
        links.push_back(line);
    }
    return links;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Использование: link_checker <файл.txt> [--threads N] [--timeout S] [--follow] [--output file]" << endl;
        return 1;
    }
    string source = argv[1];
    int threads = 4;
    int timeout_sec = 5;
    bool follow = false;
    string output;
    for (int i = 2; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "--threads" && i+1 < argc) {
            threads = stoi(argv[++i]);
        } else if (arg == "--timeout" && i+1 < argc) {
            timeout_sec = stoi(argv[++i]);
        } else if (arg == "--follow") {
            follow = true;
        } else if (arg == "--output" && i+1 < argc) {
            output = argv[++i];
        }
    }

    curl_global_init(CURL_GLOBAL_DEFAULT);

    vector<string> links;
    if (source.find(".txt") != string::npos) {
        links = load_links(source);
    } else {
        links.push_back(source);
        for (int i = 2; i < argc; ++i) {
            if (string(argv[i])[0] != '-')
                links.push_back(argv[i]);
        }
    }

    if (links.empty()) {
        cerr << "Нет ссылок для проверки." << endl;
        return 1;
    }

    cout << "Проверка " << links.size() << " ссылок (потоков: " << threads << ", таймаут: " << timeout_sec << "с)..." << endl;
    auto start = chrono::steady_clock::now();

    vector<Result> results;
    mutex results_mutex;
    vector<thread> workers;
    int chunk = (links.size() + threads - 1) / threads;
    for (int t = 0; t < threads && t * chunk < (int)links.size(); ++t) {
        int begin = t * chunk;
        int end = min(begin + chunk, (int)links.size());
        workers.emplace_back([&, begin, end]() {
            for (int i = begin; i < end; ++i) {
                string url = links[i];
                long status = check_url(url, timeout_sec, follow);
                lock_guard<mutex> lock(results_mutex);
                results.push_back({url, status, status == 0});
            }
        });
    }
    for (auto& t : workers) t.join();

    auto elapsed = chrono::duration<double>(chrono::steady_clock::now() - start).count();

    cout << "\nРезультаты:" << endl;
    for (const auto& r : results) {
        string color;
        if (r.status >= 200 && r.status < 300) color = "\033[92m";
        else if (r.status >= 300 && r.status < 400) color = "\033[93m";
        else if (r.status >= 400 && r.status < 600) color = "\033[91m";
        else color = "\033[90m";
        cout << "  " << r.url << " -> " << color << r.status << "\033[0m" << endl;
    }

    int total = results.size();
    int ok = 0, redirect = 0, error = 0, fail = 0;
    for (auto& r : results) {
        if (r.status == 0) fail++;
        else if (r.status >= 200 && r.status < 300) ok++;
        else if (r.status >= 300 && r.status < 400) redirect++;
        else if (r.status >= 400 && r.status < 600) error++;
    }
    cout << "\nСтатистика: Всего: " << total << ", OK: " << ok << ", Редиректы: " << redirect
         << ", Ошибки: " << error << ", Сбои: " << fail << endl;
    cout << "Время: " << elapsed << " сек." << endl;

    if (!output.empty()) {
        // экспорт (упрощённо)
        cout << "Экспорт в " << output << " (не реализован для краткости)" << endl;
    }

    curl_global_cleanup();
    return 0;
}

File: link_checker.cs
// link_checker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using System.Threading;

class LinkChecker
{
    private static readonly string RESET = "\x1b[0m";
    private static readonly string GREEN = "\x1b[92m";
    private static readonly string YELLOW = "\x1b[93m";
    private static readonly string RED = "\x1b[91m";
    private static readonly string GRAY = "\x1b[90m";

    static async Task<int> CheckUrl(HttpClient client, string url, int timeoutSec, bool follow)
    {
        try
        {
            var cts = new CancellationTokenSource(TimeSpan.FromSeconds(timeoutSec));
            var request = new HttpRequestMessage(HttpMethod.Get, url);
            request.Headers.UserAgent.TryParseAdd("LinkChecker/1.0");
            var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cts.Token);
            if (!follow && (response.StatusCode == System.Net.HttpStatusCode.Redirect ||
                            response.StatusCode == System.Net.HttpStatusCode.MovedPermanently ||
                            response.StatusCode == System.Net.HttpStatusCode.Found ||
                            response.StatusCode == System.Net.HttpStatusCode.SeeOther ||
                            response.StatusCode == System.Net.HttpStatusCode.TemporaryRedirect ||
                            response.StatusCode == System.Net.HttpStatusCode.PermanentRedirect))
            {
                // возвращаем статус (останавливаемся)
                return (int)response.StatusCode;
            }
            return (int)response.StatusCode;
        }
        catch
        {
            return -1;
        }
    }

    static List<string> LoadLinks(string filename)
    {
        var lines = File.ReadAllLines(filename);
        return lines.Where(l => !string.IsNullOrWhiteSpace(l) && !l.StartsWith("#")).Select(l => l.Trim()).ToList();
    }

    static async Task Main(string[] args)
    {
        if (args.Length == 0)
        {
            Console.Error.WriteLine("Использование: dotnet run <файл.txt> [--threads N] [--timeout S] [--follow] [--output file]");
            return;
        }
        string source = args[0];
        int threads = 4;
        int timeoutSec = 5;
        bool follow = false;
        string output = null;
        for (int i = 1; i < args.Length; i++)
        {
            if (args[i] == "--threads" && i+1 < args.Length)
                threads = int.Parse(args[++i]);
            else if (args[i] == "--timeout" && i+1 < args.Length)
                timeoutSec = int.Parse(args[++i]);
            else if (args[i] == "--follow")
                follow = true;
            else if (args[i] == "--output" && i+1 < args.Length)
                output = args[++i];
        }

        List<string> links;
        if (source.EndsWith(".txt"))
            links = LoadLinks(source);
        else
            links = args.ToList();

        if (!links.Any())
        {
            Console.Error.WriteLine("Нет ссылок для проверки.");
            return;
        }

        Console.WriteLine($"Проверка {links.Count} ссылок (потоков: {threads}, таймаут: {timeoutSec}с)...");
        var start = DateTime.Now;

        var results = new Dictionary<string, int>();
        var httpClient = new HttpClient();
        var semaphore = new SemaphoreSlim(threads);
        var tasks = new List<Task>();

        foreach (var url in links)
        {
            await semaphore.WaitAsync();
            tasks.Add(Task.Run(async () =>
            {
                try
                {
                    int status = await CheckUrl(httpClient, url, timeoutSec, follow);
                    lock (results) results[url] = status;
                }
                finally
                {
                    semaphore.Release();
                }
            }));
        }
        await Task.WhenAll(tasks);

        var elapsed = DateTime.Now - start;

        Console.WriteLine("\nРезультаты:");
        foreach (var url in links)
        {
            int status = results.GetValueOrDefault(url, -1);
            string color;
            if (status == -1) color = GRAY;
            else if (status >= 200 && status < 300) color = GREEN;
            else if (status >= 300 && status < 400) color = YELLOW;
            else if (status >= 400 && status < 600) color = RED;
            else color = GRAY;
            Console.WriteLine($"  {url} -> {color}{status}{RESET}");
        }

        int total = results.Count;
        int ok = results.Values.Count(s => s >= 200 && s < 300);
        int redirect = results.Values.Count(s => s >= 300 && s < 400);
        int error = results.Values.Count(s => s >= 400 && s < 600);
        int fail = results.Values.Count(s => s == -1);
        Console.WriteLine($"\nСтатистика: Всего: {total}, OK: {ok}, Редиректы: {redirect}, Ошибки: {error}, Сбои: {fail}");
        Console.WriteLine($"Время: {elapsed.TotalSeconds:F2} сек.");

        if (output != null)
        {
            Console.WriteLine($"Экспорт в {output} (не реализован для краткости)");
        }
    }
}
